File: router/src/classifier.ts
/**
 * Task complexity classifier.
 * Scores tasks 1-10 based on content analysis to determine model routing.
 * @since 2026-04
 */

export interface ClassificationResult {
  complexity: number; // 1-10
  category: string;
  reasoning: string;
}

const COMPLEXITY_SIGNALS = {
  // High complexity signals (8-10)
  high: [
    /architect/i,
    /design\s+(system|api|schema|database)/i,
    /tradeoff/i,
    /trade-off/i,
    /complex\s+reasoning/i,
    /multi.?step\s+plan/i,
    /security\s+(audit|review|analysis)/i,
    /refactor\s+(entire|whole|complete)/i,
    /migration\s+strateg/i,
  ],
  // Medium-high complexity (6-7)
  mediumHigh: [
    /debug/i,
    /code\s+review/i,
    /implement.*integrat/i,
    /fix.*bug/i,
    /performance\s+(optim|improv|tun)/i,
    /test.*coverage/i,
    /error\s+handling/i,
  ],
  // Medium complexity (4-5)
  medium: [
    /implement/i,
    /write\s+(code|function|class|module)/i,
    /create\s+(endpoint|api|component)/i,
    /add\s+(feature|functionality)/i,
    /edit\s+(file|code)/i,
    /update/i,
    /coding/i,
  ],
  // Low complexity (2-3)
  low: [
    /summarize/i,
    /explain/i,
    /describe/i,
    /list/i,
    /format/i,
    /convert/i,
    /translate/i,
    /general/i,
  ],
  // Trivial (1)
  trivial: [
    /classify/i,
    /yes\s+or\s+no/i,
    /true\s+or\s+false/i,
    /which\s+(one|option)/i,
    /quick\s+question/i,
    /lookup/i,
  ],
};

export function classifyTask(prompt: string, hint?: number): ClassificationResult {
  if (hint !== undefined && Number.isInteger(hint) && hint >= 1 && hint <= 10) {
    return {
      complexity: hint,
      category: getCategoryForComplexity(hint),
      reasoning: `User-provided complexity hint: ${hint}`,
    };
  }

  let score = 5; // Default to medium
  const matches: string[] = [];

  for (const [level, patterns] of Object.entries(COMPLEXITY_SIGNALS)) {
    for (const pattern of patterns) {
      if (pattern.test(prompt)) {
        matches.push(`${level}: ${pattern.source}`);
      }
    }
  }

  // Count matches at each level
  const highMatches = matches.filter((m) => m.startsWith("high")).length;
  const mediumHighMatches = matches.filter((m) => m.startsWith("mediumHigh")).length;
  const mediumMatches = matches.filter((m) => m.startsWith("medium")).length;
  const lowMatches = matches.filter((m) => m.startsWith("low")).length;
  const trivialMatches = matches.filter((m) => m.startsWith("trivial")).length;

  if (highMatches > 0) score = 9 + Math.min(highMatches - 1, 1);
  else if (mediumHighMatches > 0) score = 7 + Math.min(mediumHighMatches - 1, 1);
  else if (mediumMatches > 0) score = 5 + Math.min(mediumMatches - 1, 1);
  else if (lowMatches > 0) score = 3 + Math.min(lowMatches - 1, 1);
  else if (trivialMatches > 0) score = 1 + Math.min(trivialMatches - 1, 1);

  // Prompt length as a secondary signal (only when no pattern matched)
  const wordCount = prompt.split(/\s+/).length;
  if (wordCount > 500) score = Math.min(score + 1, 10);
  if (wordCount < 10 && matches.length === 0) score = Math.max(score - 1, 1);

  score = Math.max(1, Math.min(10, score));

  return {
    complexity: score,
    category: getCategoryForComplexity(score),
    reasoning: matches.length > 0
      ? `Matched signals: ${matches.join(", ")}`
      : `Default classification based on prompt length (${wordCount} words)`,
  };
}

/** Map router complexity (1-10) to workflow team size (1-3).
 *  1-4 → 1 coder + reviewer, 5-7 → 2 coders + reviewer, 8-10 → full roster. */
export function complexityToTeamSize(complexity: number): number {
  if (complexity <= 4) return 1;
  if (complexity <= 7) return 2;
  return 3;
}

function getCategoryForComplexity(c: number): string {
  if (c >= 9) return "architecture";
  if (c >= 7) return "complex-coding";
  if (c >= 5) return "coding";
  if (c >= 3) return "general";
  return "trivial";
}

File: lib/crucible/router/classifier.ex
defmodule Crucible.Router.Classifier do
  @moduledoc """
  Maps an LLM task to a complexity score 1–10.

  1-2  — trivial lookups, classification
  3-4  — general summaries, simple transforms
  5-6  — coding, moderate synthesis
  7-8  — debugging, complex coding, code review
  9-10 — architecture, tradeoff analysis, cross-system design

  The scoring combines an optional user-provided hint with regex-based
  signal matching on the prompt text. The regexes are the same set as the
  upstream TS classifier; keep them aligned when the TS changes.
  """

  @type result :: %{
          complexity: pos_integer(),
          category: String.t(),
          reasoning: String.t()
        }

  @high [
    ~r/architect/i,
    ~r/design\s+(system|api|schema|database)/i,
    ~r/tradeoff/i,
    ~r/trade-off/i,
    ~r/complex\s+reasoning/i,
    ~r/multi.?step\s+plan/i,
    ~r/security\s+(audit|review|analysis)/i,
    ~r/refactor\s+(entire|whole|complete)/i,
    ~r/migration\s+strateg/i
  ]

  @medium_high [
    ~r/debug/i,
    ~r/code\s+review/i,
    ~r/implement.*integrat/i,
    ~r/fix.*bug/i,
    ~r/performance\s+(optim|improv|tun)/i,
    ~r/test.*coverage/i,
    ~r/error\s+handling/i
  ]

  @medium [
    ~r/implement/i,
    ~r/write\s+(code|function|class|module)/i,
    ~r/create\s+(endpoint|api|component)/i,
    ~r/add\s+(feature|functionality)/i,
    ~r/edit\s+(file|code)/i,
    ~r/update/i,
    ~r/coding/i
  ]

  @low [
    ~r/summarize/i,
    ~r/explain/i,
    ~r/describe/i,
    ~r/list/i,
    ~r/format/i,
    ~r/convert/i,
    ~r/translate/i,
    ~r/general/i
  ]

  @trivial [
    ~r/classify/i,
    ~r/yes\s+or\s+no/i,
    ~r/true\s+or\s+false/i,
    ~r/which\s+(one|option)/i,
    ~r/quick\s+question/i,
    ~r/lookup/i
  ]

  @doc "Classify a prompt. If `hint` is in 1..10, it short-circuits the regex analysis."
  @spec classify(String.t(), pos_integer() | nil) :: result()
  def classify(prompt, hint \\ nil) do
    case hint do
      n when is_integer(n) and n >= 1 and n <= 10 ->
        %{complexity: n, category: category(n), reasoning: "User-provided complexity hint: #{n}"}

      _ ->
        score_from_prompt(prompt)
    end
  end

  defp score_from_prompt(prompt) when is_binary(prompt) do
    high = count_matches(prompt, @high)
    mhigh = count_matches(prompt, @medium_high)
    med = count_matches(prompt, @medium)
    low_ = count_matches(prompt, @low)
    triv = count_matches(prompt, @trivial)

    {score, reasoning} =
      cond do
        high > 0 -> {9 + min(high - 1, 1), "Matched high-complexity signals"}
        mhigh > 0 -> {7 + min(mhigh - 1, 1), "Matched medium-high signals"}
        med > 0 -> {5 + min(med - 1, 1), "Matched medium signals"}
        low_ > 0 -> {3 + min(low_ - 1, 1), "Matched low signals"}
        triv > 0 -> {1 + min(triv - 1, 1), "Matched trivial signals"}
        true -> {5, "No signals; default"}
      end

    # Prompt-length heuristic (only when we hadn't matched anything).
    word_count = prompt |> String.split() |> length()

    score =
      cond do
        high + mhigh + med + low_ + triv == 0 and word_count < 10 -> max(score - 1, 1)
        word_count > 500 -> min(score + 1, 10)
        true -> score
      end

    %{complexity: score, category: category(score), reasoning: reasoning}
  end

  defp score_from_prompt(_),
    do: %{complexity: 5, category: "coding", reasoning: "Non-string prompt"}

  defp count_matches(prompt, regexes),
    do: Enum.count(regexes, &Regex.match?(&1, prompt))

  defp category(c) when c >= 9, do: "architecture"
  defp category(c) when c >= 7, do: "complex-coding"
  defp category(c) when c >= 5, do: "coding"
  defp category(c) when c >= 3, do: "general"
  defp category(_), do: "trivial"
end
